File: python/程序员面试金典/面试题 04.06. 后继者.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def inorderSuccessor(self, root: TreeNode, p: TreeNode) -> TreeNode:
        if not root or not p: return
        if root.val <= p.val:
            return self.inorderSuccessor(root.right,p)
        else:
            left = self.inorderSuccessor(root.left,p)
            return left if left else root

File: go/程序员面试金典/面试题 04.06. 后继者.go
package main

func inorderSuccessor(root *TreeNode, p *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}
	if root.Val <= p.Val {
		return inorderSuccessor(root.Right, p)
	} else {
		left := inorderSuccessor(root.Left, p)
		if left != nil {
			return left
		} else {
			return root
		}
	}
}
